### PR TITLE
Use simplification rules also for IULIA

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libevmasm/RuleList.h [hunk 1]
```diff
@@ -43,90 +45,91 @@ template <class S> S modWorkaround(S const& _a, S const& _b)
...
-		{{Instruction::XOR, {X, X}}, [=]{ return u256(0); }},
-		{{Instruction::SUB, {X, X}}, [=]{ return u256(0); }},
-		{{Instruction::EQ, {X, X}}, [=]{ return u256(1); }},
-		{{Instruction::LT, {X, X}}, [=]{ return u256(0); }},
-		{{Instruction::SLT, {X, X}}, [=]{ return u256(0); }},
-		{{Instruction::GT, {X, X}}, [=]{ return u256(0); }},
-		{{Instruction::SGT, {X, X}}, [=]{ return u256(0); }},
-		{{Instruction::MOD, {X, X}}, [=]{ return u256(0); }},
+		{{Instruction::AND, {X, X}}, [=]{ return X; }, true},
+		{{Instruction::OR, {X, X}}, [=]{ return X; }, true},
```
Comment (axic) [inline libevmasm/RuleList.h]: Isn't this false?

Comment (axic) [reply libevmasm/RuleList.h]: Ok, if it has side effect for executing twice, then this is correct.

File: libjulia/optimiser/SimplificationRules.h [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	/// Inside one rule, all patterns in the same match group have to match expressions from the
+	/// same expression equivalence class.
+	void setMatchGroup(unsigned _group, std::map<unsigned, Expression const*>& _matchGroups);
+	unsigned matchGroup() const { return m_matchGroup; }
+	bool matches(Expression const& _expr) const;
+
+	std::vector<Pattern> arguments() const { return m_arguments; }
+
+	/// @returns the data of the matched expression if this pattern is part of a match group.
+	u256 d() const;
```
Comment (axic) [inline libjulia/optimiser/SimplificationRules.h:97]: Is this used?

Comment (chriseth) [reply libjulia/optimiser/SimplificationRules.h:97]: Yes, this is the main thing used in the pattern action part.

File: libjulia/optimiser/SyntacticalEquality.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		return boost::get<Identifier>(_e1).name == boost::get<Identifier>(_e2).name;
+	else if (_e1.type() == typeid(Literal))
+	{
+		auto const& e1 = boost::get<Literal>(_e1);
+		auto const& e2 = boost::get<Literal>(_e2);
+		return e1.kind == e2.kind && e1.value == e2.value && e1.type == e2.type;
+	}
+	else
+	{
+		solAssert(false, "Invlid expression");
```
Comment (axic) [inline libjulia/optimiser/SyntacticalEquality.cpp:65]: typo

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Fails on travis:

clang:
```
/home/travis/build/ethereum/solidity/libjulia/optimiser/SimplificationRules.h:45:42: error: expected class name
class SimplificationRules: public boost::noncopyable
```

gcc:
```
In file included from /home/travis/build/ethereum/solidity/libjulia/optimiser/ExpressionSimplifier.cpp:23:0:
/home/travis/build/ethereum/solidity/libjulia/optimiser/SimplificationRules.h:46:1: error: expected class-name before ‘{’ token
 {
 ^
```

**Comment by chriseth** [comment]
Travis had a stack size failure while compiling emscripten. Triggered again. I hope this does not mean that our code got too large...

**Comment by chriseth** [comment]
Ok, now it's only the zeppelin tests.

Please confirm again, @axic, that this can be merged.

File: scripts/travis-emscripten/build_emscripten.sh [hunk 4]
```diff
@@ -42,6 +42,10 @@ fi
 
 WORKSPACE=/root/project
 
+# Increase nodejs stack size
+sed -i -e 's/NODE_JS="nodejs"/NODE_JS=["nodejs", "--stack_size=8192"]/' /root/.emscripten
```
Comment (axic) [inline scripts/travis-emscripten/build_emscripten.sh]: I'm not familiar with the emscripten compilation process but what is `node` used for there and what was failing without the stack size increase?

Comment (axic) [reply scripts/travis-emscripten/build_emscripten.sh]: For the record: https://travis-ci.org/ethereum/solidity/jobs/338054884

```
Linking CXX executable soljson.js
/emsdk_portable/emscripten/tag-1.35.4/tools/js-optimizer.js:7758
    print(JSON.stringify(ast));
               ^
RangeError: Maximum call stack size exceeded
fs.js:436
  return binding.open(pathModule._makeLong(path), stringToFlags(flags), mode);
                 ^
Error: ENOENT, no such file or directory '/tmp/tmpEOtDSr.jsfunc_7.js'
    at Object.fs.openSync (fs.js:436:18)
    at Object.fs.readFileSync (fs.js:290:15)
    at read (/emsdk_portable/emscripten/tag-1.35.4/tools/js-optimizer.js:68:34)
    at Object.<anonymous> (/emsdk_portable/emscripten/tag-1.35.4/tools/js-optimizer.js:7708:11)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
```

So it seems the output code AST became too big, but at least the resulting .js file is still only 7.5MB.

**Comment by axic** [comment]
It breaks circleci:

```
# Increase nodejs stack size
sed -i -e 's/NODE_JS="nodejs"/NODE_JS=["nodejs", "--stack_size=8192"]/' /root/.emscripten
sed: can't read /root/.emscripten: No such file or directory
Exited with code 2
```

 :cry:

File: scripts/travis-emscripten/build_emscripten.sh (hunk 4 above)
Comment (chriseth) [reply scripts/travis-emscripten/build_emscripten.sh]: It looks like emscripten has an optimizer written in javascript which is run via node.

**Comment by chriseth** [comment]
Ah, so many complications in the simplifier ;-)

**Comment by chriseth** [comment]
Ah, travis and circle use a different emscripten docker image version!

**Comment by chriseth** [comment]
Failure on circle is just zeppelin test.

File: scripts/travis-emscripten/build_emscripten.sh [hunk 5]
```diff
@@ -43,7 +43,12 @@ fi
 WORKSPACE=/root/project
 
 # Increase nodejs stack size
-sed -i -e 's/NODE_JS="nodejs"/NODE_JS=["nodejs", "--stack_size=8192"]/' /root/.emscripten
+if [ -e ~/.emscripten ]
+then
+    sed -i -e 's/NODE_JS="nodejs"/NODE_JS=["nodejs", "--stack_size=8192"]/' ~/.emscripten
```
Comment (axic) [inline scripts/travis-emscripten/build_emscripten.sh]: Should have comments explaining why there is a difference. Also would have squashed the commit :)

Comment (chriseth) [reply scripts/travis-emscripten/build_emscripten.sh]: Oh yeah, you are right...

**Pull request merged**